// @azure/cosmos/lib/client/Item/Item.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const common_1 = require("../../common");
/**
 * Used to perform operations on a specific item.
 *
 * @see {@link Items} for operations on all items; see `container.items`.
 */
class Item {
    /**
     * @hidden
     * @param container The parent {@link Container}.
     * @param id The id of the given {@link Item}.
     * @param primaryKey The primary key of the given {@link Item} (only for partitioned containers).
     */
    constructor(container, id, primaryKey, clientContext) {
        this.container = container;
        this.id = id;
        this.primaryKey = primaryKey;
        this.clientContext = clientContext;
    }
    /**
     * Returns a reference URL to the resource. Used for linking in Permissions.
     */
    get url() {
        return common_1.UriFactory.createDocumentUri(this.container.database.id, this.container.id, this.id);
    }
    read(options) {
        return __awaiter(this, void 0, void 0, function* () {
            options = options || {};
            if ((!options || !options.partitionKey) && this.primaryKey) {
                options.partitionKey = this.primaryKey;
            }
            const path = common_1.Helper.getPathFromLink(this.url);
            const id = common_1.Helper.getIdFromLink(this.url);
            const response = yield this.clientContext.read(path, "docs", id, undefined, options);
            return {
                body: response.result,
                headers: response.headers,
                ref: this,
                item: this
            };
        });
    }
    replace(body, options) {
        return __awaiter(this, void 0, void 0, function* () {
            options = options || {};
            if ((!options || !options.partitionKey) && this.primaryKey) {
                options.partitionKey = this.primaryKey;
            }
            if (options.partitionKey === undefined && options.skipGetPartitionKeyDefinition !== true) {
                const { body: partitionKeyDefinition } = yield this.container.getPartitionKeyDefinition();
                options.partitionKey = this.container.extractPartitionKey(body, partitionKeyDefinition);
            }
            const err = {};
            if (!common_1.Helper.isResourceValid(body, err)) {
                throw err;
            }
            const path = common_1.Helper.getPathFromLink(this.url);
            const id = common_1.Helper.getIdFromLink(this.url);
            const response = yield this.clientContext.replace(body, path, "docs", id, undefined, options);
            return {
                body: response.result,
                headers: response.headers,
                ref: this,
                item: this
            };
        });
    }
    delete(options) {
        return __awaiter(this, void 0, void 0, function* () {
            options = options || {};
            if ((!options || !options.partitionKey) && this.primaryKey) {
                options.partitionKey = this.primaryKey;
            }
            const path = common_1.Helper.getPathFromLink(this.url);
            const id = common_1.Helper.getIdFromLink(this.url);
            const response = yield this.clientContext.delete(path, "docs", id, undefined, options);
            return {
                body: response.result,
                headers: response.headers,
                ref: this,
                item: this
            };
        });
    }
}
exports.Item = Item;
