// @azure/cosmos/lib/queryExecutionContext/EndpointComponent/TopEndpointComponent.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
/** @hidden */
class TopEndpointComponent {
    /**
     * Represents an endpoint in handling top query. It only returns as many results as top arg specified.
     * @constructor TopEndpointComponent
     * @param { object } executionContext - Underlying Execution Context
     * @ignore
     */
    constructor(executionContext, topCount) {
        this.executionContext = executionContext;
        this.topCount = topCount;
    }
    /**
     * Execute a provided function on the next element in the TopEndpointComponent.
     * @memberof TopEndpointComponent
     * @instance
     * @param {callback} callback - Function to execute for each element. \
     * the function takes two parameters error, element.
     */
    nextItem() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.topCount <= 0) {
                return { result: undefined, headers: undefined };
            }
            this.topCount--;
            try {
                return this.executionContext.nextItem();
            }
            catch (err) {
                throw err;
            }
        });
    }
    /**
     * Retrieve the current element on the TopEndpointComponent.
     * @memberof TopEndpointComponent
     * @instance
     * @param {callback} callback - Function to execute for the current element. \
     * the function takes two parameters error, element.
     */
    current() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.topCount <= 0) {
                return { result: undefined, headers: undefined };
            }
            try {
                return this.executionContext.current();
            }
            catch (err) {
                throw err;
            }
        });
    }
    /**
     * Determine if there are still remaining resources to processs.
     * @memberof TopEndpointComponent
     * @instance
     * @returns {Boolean} true if there is other elements to process in the TopEndpointComponent.
     */
    hasMoreResults() {
        return this.topCount > 0 && this.executionContext.hasMoreResults();
    }
}
exports.TopEndpointComponent = TopEndpointComponent;
