// @azure/cosmos/lib/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
const DocumentBase = require("./documents");
exports.DocumentBase = DocumentBase;
exports.AzureDocuments = DocumentBase;
var documents_1 = require("./documents");
exports.ConnectionMode = documents_1.ConnectionMode;
exports.ConsistencyLevel = documents_1.ConsistencyLevel;
exports.ConnectionPolicy = documents_1.ConnectionPolicy;
exports.DatabaseAccount = documents_1.DatabaseAccount;
exports.DataType = documents_1.DataType;
exports.IndexingMode = documents_1.IndexingMode;
exports.IndexKind = documents_1.IndexKind;
exports.MediaReadMode = documents_1.MediaReadMode;
exports.PartitionKind = documents_1.PartitionKind;
exports.PermissionMode = documents_1.PermissionMode;
exports.QueryCompatibilityMode = documents_1.QueryCompatibilityMode;
exports.TriggerOperation = documents_1.TriggerOperation;
exports.TriggerType = documents_1.TriggerType;
exports.UserDefinedFunctionType = documents_1.UserDefinedFunctionType;
var common_1 = require("./common");
exports.Constants = common_1.Constants;
exports.UriFactory = common_1.UriFactory;
var retry_1 = require("./retry");
exports.RetryOptions = retry_1.RetryOptions;
var queryIterator_1 = require("./queryIterator");
exports.QueryIterator = queryIterator_1.QueryIterator;
__export(require("./queryMetrics"));
var CosmosClient_1 = require("./CosmosClient");
exports.CosmosClient = CosmosClient_1.CosmosClient;
__export(require("./client"));
